### PR TITLE
Ensure order of parents of Intercom articles for Vault

### LINKED ISSUE #6945: [Vaults Infra] Double check that Intercom follow parent convention
The current standard for the parents array in a Document is set as [currentElement, ..., n], where n represents the oldest ancestor. Can we double check that Intercom help centers and conversations are properly implementing this behavior.

### PR DESCRIPTION
**Pull request by PopDaph**
## Description

Fixes https://github.com/dust-tt/dust/issues/6945
Ensure Intercom articles have their parents in order since it will be a requirement for vaults.

_// Need to fix my local intercom before merging cause this is untested, but opening the PR already to get a review._

## Risk

Break intercom. 

## Deploy Plan

Deploy connector. 
Full resync of  all help centers?

### CODE REVIEW

File: connectors/src/connectors/intercom/temporal/sync_help_center.ts [hunk 1]
```diff
@@ -367,16 +369,56 @@ export async function upsertArticle({
...
-      getHelpCenterCollectionInternalId(connectorId, id.toString())
+    // Order is important: We want the id of the article, then all parents collection in order, then the help center
+    const articleinternalId = documentId;
+    const parentCollectioninternalId = getHelpCenterCollectionInternalId(
+      connectorId,
+      parentCollectionId
     );
-    parentsInternalsIds.push(
-      getHelpCenterInternalId(connectorId, helpCenterId)
+    let grandParentCollectioninternalId = null;
```
Comment (flvndvd) [inline connectors/src/connectors/intercom/temporal/sync_help_center.ts:381]:
```suggestion
    let grandParentCollectionInternalId = null;
```

File: connectors/src/connectors/intercom/temporal/sync_help_center.ts [hunk 2]
```diff
@@ -367,16 +369,56 @@ export async function upsertArticle({
...
+    let greatGrandParentCollectioninternalId = null;
```
Comment (flvndvd) [inline connectors/src/connectors/intercom/temporal/sync_help_center.ts:382]:
```suggestion
    let greatGrandParentCollectionInternalId = null;
```

File: connectors/src/connectors/intercom/temporal/sync_help_center.ts [hunk 3]
```diff
@@ -367,16 +369,56 @@ export async function upsertArticle({
...
+    const helpCenterinternalId = getHelpCenterInternalId(
+      connectorId,
+      helpCenterId
     );
-    const documentId = getHelpCenterArticleInternalId(connectorId, article.id);
-    parentsInternalsIds.push(documentId);
+
+    // Then we will check for grand-parents collections. There's max 2-level on Intercom.
```
Comment (flvndvd) [inline connectors/src/connectors/intercom/temporal/sync_help_center.ts:388]:
```suggestion
    // Then we will check for grand-parents collections. There's max 2-levels on Intercom.
```

File: connectors/src/connectors/intercom/temporal/sync_help_center.ts (hunk 1 above)
Comment (flvndvd) [inline connectors/src/connectors/intercom/temporal/sync_help_center.ts:381]: Suggestion to improve code readability, a new function something like 👇 . I guess you'll also need it for the backfill anyway.

```
// Parents in the Core datasource should map the internal ids that we use in the permission modal
// Order is important: We want the id of the article, then all parents collection in order, then the help center
async function getParentIdsForArticle(documentId, connectorId, parentCollectionId, helpCenterId) {
  // Initialize the internal IDs array with the article ID.
  const internalIds = [documentId];

  // Add the parent collection ID.
  internalIds.push(getHelpCenterCollectionInternalId(connectorId, parentCollectionId));

  // Fetch and add any grandparent collection IDs.
  let currentParentId = parentCollectionId;
  
  // There's max 2-levels on Intercom.
  for (let i = 0; i < 2; i++) { 
    const currentParent = await IntercomCollection.findOne({
      where: {
        connectorId,
        collectionId: currentParentId,
      },
    });

    if (currentParent && currentParent.parentId) {
      currentParentId = currentParent.parentId;
      internalIds.push(getHelpCenterCollectionInternalId(connectorId, currentParentId));
    } else {
      break;
    }
  }

  // Add the help center internal ID
  internalIds.push(getHelpCenterInternalId(connectorId, helpCenterId));

  return internalIds;
}
```

Comment (PopDaph) [reply connectors/src/connectors/intercom/temporal/sync_help_center.ts:381]: I had something like that initially and those this is less readable 😅

**Review by flvndvd — APPROVED** [review]
LGTM. What's the plan to backfill?

**Comment by PopDaph** [comment]
https://github.com/dust-tt/dust/pull/6960

**Pull request merged**